src/renderer/atlas/AtlasEngine.h

```diff
@@ -0,0 +1,761 @@
+
+            static constexpr size_t dataSize(u16 charCount) noexcept
+            {
+                // This returns the actual byte size of a AtlasKeyData struct for the given charCount.
+                // The `wchar_t chars[2]` is only a buffer for the inlined variant after
+                // all and the actual charCount can be smaller or larger. Due to this we
+                // remove the size of the `chars` array and add it's true length on top.
+                return sizeof(AtlasKeyData) - sizeof(AtlasKeyData::chars) + static_cast<size_t>(charCount) * sizeof(AtlasKeyData::chars[0]);
+            }
+        };
+
+        struct AtlasKeyHasher
+        {
+            size_t operator()(const AtlasKey& key) const noexcept
+            {
+                return key.hash();
+            }
+        };
+
+        struct AtlasValueData
+        {
+            CellFlags flags = CellFlags::None;
+            u16x2 coords[7];
+        };
+
+        struct AtlasValue
+        {
+            constexpr AtlasValue() = default;
+
+            u16x2* initialize(CellFlags flags, u16 cellCount)
+            {
+                const auto size = dataSize(cellCount);
+                const auto data = _data.initialize(size);
+                WI_SetFlagIf(flags, CellFlags::Inlined, _data.would_inline(size));
+                data->flags = flags;
+                return &data->coords[0];
+            }
+
+            const AtlasValueData* data() const noexcept
+            {
+                return _data.data();
+            }
+
+        private:
+            SmallObjectOptimizer<AtlasValueData> _data;
+
+            static constexpr size_t dataSize(u16 coordCount) noexcept
+            {
+                return sizeof(AtlasValueData) - sizeof(AtlasValueData::coords) + static_cast<size_t>(coordCount) * sizeof(AtlasValueData::coords[0]);
+            }
+        };
+
+        struct AtlasQueueItem
+        {
+            const AtlasKey* key;
+            const AtlasValue* value;
+            float scale;
+        };
+
+        struct CachedCursorOptions
+        {
+            u32 cursorColor = INVALID_COLOR;
+            u16 cursorType = gsl::narrow_cast<u16>(CursorType::Legacy);
+            u8 heightPercentage = 20;
+
+            ATLAS_POD_OPS(CachedCursorOptions)
+        };
+
+        struct BufferLineMetadata
+        {
+            u32x2 colors;
+            CellFlags flags = CellFlags::None;
+        };
+
+        // NOTE: D3D constant buffers sizes must be a multiple of 16 bytes.
+        struct alignas(16) ConstBuffer
+        {
+            // WARNING: Modify this carefully after understanding how HLSL struct packing works.
+            // The gist is:
+            // * Minimum alignment is 4 bytes (like `#pragma pack 4`)
+            // * Members cannot straddle 16 byte boundaries
+            //   This means a structure like {u32; u32; u32; u32x2} would require
+            //   padding so that it is {u32; u32; u32; <4 byte padding>; u32x2}.

```

Comment:
can the job here be done with `alignas`?

Reply:
It is! The way this code is written, as long as others adhere to this style, it'll produce compilation errors if you violate this warning.